doc/mgmt/gnmi/SONiC_GNMI_Server_Interface_Design.md

```diff
@@ -0,0 +1,568 @@
+| Generic Config Updater | SONiC configuration generic update and rollback feature |
+| ConfigDB | Configuration Database |
+| ApplDB | Application Database |
+
+# 1 Feature Overview
+
+SONiC should provide a gNMI/gNOI server interface where the client can communicate with SONiC (server) through the interface using their respective gRPC libraries based on ConfigDB/ApplDB/StateDB/CountersDB schema or SONiC Yang schema.
+
+To understand the requirements, we need to go over current SONiC management components:
+
+* SONIC CLI commands
+
+<img src="images/cli.svg" alt="cli" width="600px"/>
+
+1. Users could use SONiC CLI command "show runningconfig all" to read current configuration. 
+    - The blue arrow indicates this process.
+2. User could use SONiC CLI command "config apply-patch" to incrementally update configurations systematically and validate target configuration using SONiC Yang models. With generic_config_updater, all operations in a 'config apply-patch' command are processed in a single transaction that will either succeed or fail as a whole.
+    - "config apply-patch" supports ConfigDB, and Yang models.
+    - The green arrow indicates this process, and dotted arrow indicates Yang validation.
+3. SONiC CLI commands can use "config reload" to erase original and fully reload configurations from a file and validate target configuration using SONiC Yang models.
+    - "config reload" already support input in ConfigDB schema or in yang models schema.
+    - The yellow arrow indicates this process, and dotted arrow indicates Yang validation.
+
+And we have the possibility to improve generic_config_updater performance.
+
+* SONIC Restapi
+
+<img src="images/restapi.svg" alt="restapi" width="800px"/>
+
+sonic-restapi is a docker container running in SONiC, which exposes HTTPS endpoints to perform dynamic configuration including config VNET routes.
+One challenge of sonic-restapi is to provide a general config schema. Currently its API is designed case by case.
+
+sonic-restapi will be replaced by gNMI interface in the future.
+
+* SONiC system telemetry services
+
+<img src="images/telemetry.png" alt="telemetry" width="500px"/>
+
+sonic-telemetry provides gNMI/gNOI server interface. And sonic-telemetry provides only telemetry via gRPC, and the gap is configuration.
+
+We will update sonic-telemetry to support configuration.
+
+## 1.1 Requirements
+
+### 1.1.1 Phase 1 Requirements
+* Set and get RPCs must be supported. Customer will use get RPC to retrieve configurations including FRR and VNET route, and use set RPC to apply new configurations.
+* SetRequest message must support to incrementally update configuration and fully update configurations.
+* Data models can be SONiC Yang models or CONFIG_DB schema.
+* Ability to read different DBs - ApplDB, ConfigDB, StateDB, CountersDB etc.
+* Ability to write different DBs - ApplDB, ConfigDB.
+* Configurations must be verified using YANG models even it is in CONFIG_DB schema.
```

Comment:
How are we planning to achieve this? Are we planning to integrate with the python validator(In which case validation of the bulking configurations will become time-consuming due to loading and unloading of pythong files)?

Reply:
For CONFIG_DB, we can accept time-consuming configurations. Validation is more important to us.

Reply:
Validation can be done by other means(go lang based validator that will not be time consuming) why integrate with python validator knowing that it is going to be time-consuming.

Reply:
We need to use generic config updater to support incremental configuration update.
For now, generic config updater is implemented with Python and it's time consuming, and we can improve generic config updater in the future.

Reply:
@anand-kumar-subramanian Could you provide some sample of "bulking configurations will become time-consuming" in your mind. If the "bulking configurations" is in one SetRequest, the "loading and unloading of python files" will happen only once and it will not be a significant performance issue. If "bulking configurations" is in multiple SetRequest, it is better to ask client to combine them into one.